Prototype build-time data loading from github

last-partizan opened the pull request:
## New feature

Hi, when searching for a library - my workflow is usually like this:

- open link, check if it's not dead, close link

If we can load repository info at build time and get data about latest commits, we can display beatiful "updated last month" text, or "not updated for a few years", or something like this.

We probably can also use `npmjs` API to check for releases, and other cool things.

Refs https://vitepress.dev/guide/data-loading

What do you think?

![image](https://github.com/rmjordas/awesome-vue/assets/301015/fc24e60f-dcee-4736-8d2c-eff24114498e)


## Contribution Agreement

By submitting a pull request to this repository, you are accepting this contribution agreement:

> I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

👌🏼

rmjordas commented:
Hello @last-partizan. Thank you for opening this pull request. I think it's a good idea to have more information about the links that are listed. And your suggestion is cool. I think it could work, but maybe we keep the number of requests done during build time under GitHub/npmjs API's rate limits.

I haven't got the time to work on this project right now BUT when Vitepress 1.0 (not RC) is released, I'll go through the whole project and remove dead links and fix typos. 👍

last-partizan commented:
Great to hear this!

I'll continue working on this idea in my free time, and think about how to efficiently migrate structure to data-oriented design, and how to stay below API limits.

last-partizan commented on README.md:
````diff
@@ -55,6 +55,15 @@ npm run build
 npm run preview --port 4173
 ```
 
+To update GitHub and npmjs stats for packages, use:
+
+```bash
+npm run fetch-data
+```
+
+When adding a package to .data.json file, skip `github` and `nodejs` fields, and
+then run `fetch-data` script.
````
@rmjordas i think, during github builds we will not access external API's

Instead, when updating page locally - we have option to run `fetch-data` for loading package stats.

Maybe, next step would be running this regularly as part of CI, but for now let's start with less risky approach.

Github allows 30 search requests per minute, and octokit library handles automatic retries.
For npm api i haven't found any details on API limits, but probably we can figure it out when we convert more pages.

rmjordas replied:
I've peeked a bit at the code and at the deploy preview and I think it looks neat. Good job 👍 I've also added you to the repository so you can work with the `next` branch however you like.

I could imagine a GitHub Action to handle regular updates for the stats. I'm not that clear right now on what the process is for adding new links, or how we'd handle submissions from external contributors, but I might need more time to study your changes to understand that.

last-partizan replied:
I was thinking about splitting the data - one json file - storing the links/descriptions, and another - with fetched stats, but decided against it, single file is simpler to work with from programming point of view.

So, we just add "name", "url" and "description" into json file - and run `fetch-data` script, it populates other fields.

Right now it will refetch stats for everything, but i can upgrade it to fetch only new data when asked.

last-partizan commented on convert-md-to-json.ts:
```diff
@@ -0,0 +1,96 @@
+/// <reference types="node" />
+/**
+ * This is the script for converting old .md files into new data-oriented scheme.
+ *
+ * It's intended to be used once, and removed after migration.
+ *
+ * When preparing migration into master, use:
+ *
+ *   git checkout master -- content/components-and-libraries/
+ *   tsx convert-md-to-json.ts
+ *   npm run fetch-data
+ *
```
Now, it's ready to be merged.

I converted all md-files into data-oriented format.

I think, we may have need this file one last time when merging into master.

@rmjordas please, take a look!

rmjordas replied:
This is great, @last-partizan! My initial thought is: Right now, this proposal would probably break the "Edit this page in GitHub" feature at the bottom of the page. 

If someone external to the project clicks this link and they wanted to add an entry, it will take them to the markdown files. However, they wouldn't be able to insert their link entry because the markdown items have been replaced by a `<ProjectList>` component.

We could remove this edit link or add more instructions for contributors or maybe something else. But first, I'm going to experiment with your proposal and figure out if the caching, conversion, and fetching are can be done while building the website. If this is the case, I'm hoping we might be able to preserve the original markdown files and there's no need to commit the `*.json` and `*.js` files to the repository. 

I know this will probably strech the build times but maybe not that much. We'll see. And also, let me know if you have solutions for this!

Give me a few days/weeks to review. 🙏

rmjordas replied:
This is great, @last-partizan! My initial thought is: Right now, this proposal would probably break the "Edit this page in GitHub" feature at the bottom of the page. 

If someone external to the project clicks this link and they wanted to add an entry, it will take them to the markdown files. However, they wouldn't be able to insert their link entry because the markdown items have been replaced by a `<ProjectList>` component.

We could remove this edit link or add more instructions for contributors or maybe something else. But first, I'm going to experiment with your proposal and figure out if the caching, conversion, and fetching are can be done while building the website. If this is the case, I'm hoping we might be able to preserve the original markdown files and there's no need to commit the `*.json` and `*.js` files to the repository. 

I know this will probably strech the build times but maybe not that much. We'll see. And also, let me know if you have solutions for this!

Give me a few days/weeks to review. 🙏

last-partizan replied:
I think, it's best to store data in structured format, because parsing it from markdown each time is more error-prone.

And, if we add clear instructions on how to add new data - it would be fine. Maybe we can change "Edit this page" link to point to data file, instead of md.

I've already found one unicode mdash instead of - :)

StefanBrand replied:
Just came here via a discussion at Vue.js Nation that mentioned Awesome Vue.

Have you considered storing data in the frontmatter? https://vitepress.dev/guide/frontmatter

last-partizan replied:
It is certainly possible to store data in the frontmatter, but i don't know what benefits of this approach.

frontmatter is better suited for storing small pieces of data, like page title and metadata. But here we trying to write data back to json files, with added remote-fetched details.

It's easier to do with separate json files.

rmjordas commented on content/.vitepress/components/ProjectList.vue:
```diff
@@ -0,0 +1,44 @@
+        <PlainTextBadge
+          v-if="item.npmjs"
+          label="at"
+          :value="new Date(item.npmjs.lastReleaseAt).toLocaleDateString()"
```
What if this shows a relative time instead, like "A few days ago"? Could mean an added dependency tho, what do you think?

rmjordas commented on content/.vitepress/components/ProjectList.vue:
```diff
@@ -0,0 +1,44 @@
+        />
+        <PlainTextBadge
+          v-if="item.npmjs"
+          label="at"
```
I wasn't sure at first glance what "at" means when viewing the deploy preview. How about "Last Released" or something like that?

rmjordas commented on content/.vitepress/components/ProjectList.vue:
```diff
@@ -0,0 +1,44 @@
+}>();
+</script>
+<style scoped>
+.badges {
```
Maybe the badges could be spaced a bit more so it's not as close to the item link.

last-partizan commented:
I found another website, which uses similar approach - with fetching data from repositories and github to add metadata for packages.

But it's more mature than this my attempt, and uses rust :)

https://github.com/anowell/are-we-learning-yet/tree/master/_scraper

(I'm not gonna rewrite this in rust, but it's nice to know where to look for ideas)

One thing we should probably borrow from this - switch from json to yaml - as it's more human-readable, and has comments.

rmjordas commented:
> https://github.com/anowell/are-we-learning-yet/tree/master/_scraper

Thanks for the link. I'll take a peek. Maybe I'll get some ideas. 🦀

last-partizan commented:
Added some padding to badges, changed a label and switched to the relative time format, turned out it doesn't even need a library!

And, after reading this: https://philna.sh/blog/2023/02/02/yaml-document-from-hell-javascript-edition/

We, probably should stick to json. Yaml has it's own problems :) But, it can be easy to switch in the future to some other data format, main point still stands - data is better than parsing markdown.

And we should parse markdown only once when converting to data-oriented format.

last-partizan commented:
I hate you, @github-actions bot!

Let's reopen this and maybe try finishing this PR.
